### PR TITLE
Codechange: Rework IniFile handling to use std::list and less pointering.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

IniFile processing uses C-style linked lists, involving manual memory management (new/delete) and linked list management.

Additionally, due to GetGroup() creating missing groups by default, reading IniFiles will temporarily create IniGroups that are not present.

IniFile processing often uses manually allocated objects instead of automatic storage.

## Description

* Add CreateItem / CreateGroup methods so that the caller does not need to know the underlying structure.
* Split GetGroup (optionally creating a group) into GetGroup and GetOrCreateGroup, as was previously done for GetItem and GetOrCreateItem.
* Replace the linked lists with std::list instead.
* Use automatic storage instead of new/delete in various places.

## Limitations

"If it ain't broken don't fix it" and "That's quite a lot of changes"

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/base_media_func.h [hunk 1]
```diff
@@ -51,10 +56,10 @@ bool BaseSet<T, Tnum_files, Tsearch_in_tars>::FillSetDetails(IniFile *ini, const
 	this->description[std::string{}] = *item->value;
 
 	/* Add the translations of the descriptions too. */
-	for (item = metadata->item; item != nullptr; item = item->next) {
-		if (item->name.compare(0, 12, "description.") != 0) continue;
+	for (IniItem &item : metadata->items) {
```
Comment (LordAro) [inline src/base_media_func.h]: Const?

File: src/settingsgen/settingsgen.cpp [hunk 2]
```diff
@@ -306,31 +292,30 @@ static void DumpLine(IniItem *item, IniGroup *grp, IniGroup *default_grp, Output
...
+	IniGroup *templates_grp = ifile.GetGroup(TEMPLATES_GROUP_NAME);
+	IniGroup *validation_grp = ifile.GetGroup(VALIDATION_GROUP_NAME);
 	if (templates_grp == nullptr) return;
 
 	/* Output every group, using its name as template name. */
-	for (IniGroup *grp = ifile->group; grp != nullptr; grp = grp->next) {
-		const char * const *sgn;
-		for (sgn = special_group_names; *sgn != nullptr; sgn++) if (grp->name == *sgn) break;
-		if (*sgn != nullptr) continue;
+	for (IniGroup &grp : ifile.groups) {
```
Comment (LordAro) [inline src/settingsgen/settingsgen.cpp]: Const?

**Review by LordAro — COMMENTED** [review]
Just guesses based on reading the diff, probably wrong

File: src/ini_type.h [hunk 3]
```diff
@@ -56,10 +56,10 @@ struct IniLoadFile {
 	IniGroup *group;                      ///< the first group in the ini
 	IniGroup **last_group;                ///< the last group in the ini
 	std::string comment;                  ///< last comment in file
-	const char * const *list_group_names; ///< nullptr terminated list with group names that are lists
-	const char * const *seq_group_names;  ///< nullptr terminated list with group names that are sequences.
+	const std::vector<const char *> list_group_names; ///< nullptr terminated list with group names that are lists
+	const std::vector<const char *> seq_group_names;  ///< nullptr terminated list with group names that are sequences.
```
Comment (rubidium42) [inline src/ini_type.h]: Are they still `nullptr terminated`?

**Comment by PeterN** [comment]
Some reorganisation happened, now uses std::initializer_list instead of std::vector (we don't want vector functionality here) but it's aliased in case that ever needs to change.

Non-const GetItem() wasn't used so that's removed.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**